File: app/dashboard/src/types/Audit.ts
export type AuditValue = Record<string, unknown> | unknown[] | string | number | boolean | null;

export interface AuditLog {
  id: number;
  admin_id: number | null;
  admin_username: string;
  action: string;
  target_type: string;
  target_id: string | null;
  target_name: string | null;
  description: string;
  previous_value: AuditValue;
  new_value: AuditValue;
  details: AuditValue;
  ip_address: string | null;
  status: "success" | "failed" | string;
  created_at: string;
}

export interface AuditLogList {
  logs: AuditLog[];
  total: number;
  offset: number;
  limit: number;
}

export interface AuditLogOptions {
  admins: string[];
  actions: string[];
}

File: app/models/audit.py
from datetime import datetime
from typing import Any, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class AuditLogResponse(BaseModel):
    id: int
    admin_id: Optional[int] = None
    admin_username: str
    action: str
    target_type: str
    target_id: Optional[str] = None
    target_name: Optional[str] = None
    description: str
    previous_value: Optional[Any] = None
    new_value: Optional[Any] = None
    details: Optional[Any] = None
    ip_address: Optional[str] = None
    status: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class AuditLogList(BaseModel):
    logs: List[AuditLogResponse]
    total: int
    offset: int
    limit: int


class AuditLogOptions(BaseModel):
    admins: List[str] = Field(default_factory=list)
    actions: List[str] = Field(default_factory=list)
